File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
public:

    vector<vector<int>> ans;
    vector<int> temp;

    void solve(int index, vector<int>& candidates, int target) {

        // Base Case
        if (target == 0) {
            ans.push_back(temp);
            return;
        }

        if (target < 0)
            return;

        for (int i = index; i < candidates.size(); i++) {

            temp.push_back(candidates[i]);

            // Same element can be used again
            solve(i, candidates, target - candidates[i]);

            // Backtrack
            temp.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        solve(0, candidates, target);

        return ans; 
    }
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        result = []

        def backtrack(start, target, path):
            # If target becomes 0, we found a valid combination
            if target == 0:
                result.append(path[:])
                return

            # If target becomes negative, stop exploring
            if target < 0:
                return

            # Try each candidate starting from 'start'
            for i in range(start, len(candidates)):
                path.append(candidates[i])

                # Reuse the same number, so pass i (not i+1)
                backtrack(i, target - candidates[i], path)

                # Backtrack
                path.pop()

        backtrack(0, target, [])
        return result
        